src/dicomweb_client/uri.py

```diff
@@ -0,0 +1,533 @@
+"""Utilities for DICOMweb URI manipulation."""
+import enum
+import re
+from typing import Optional, Sequence, Tuple
+import urllib.parse as urlparse
+
+
+class URIType(enum.Enum):
+    """Type of DICOM resource the URI points to."""
+    SERVICE = 'service'
+    STUDY = 'study'
+    SERIES = 'series'
+    INSTANCE = 'instance'
+    FRAME = 'frame'
+
+
+class URISuffix(enum.Enum):
+    """Optional suffixes for a DICOM resource."""
+    METADATA = 'metadata'
+    RENDERED = 'rendered'
+    THUMBNAIL = 'thumbnail'
+
+
+# For DICOM Standard spec validation of UID components in `URI`.
+_MAX_UID_LENGTH = 64
+_REGEX_UID = re.compile(r'[0-9]+([.][0-9]+)*')
+
+
+class URI:
```

Comment:
Thank you for the thoughtful feedback. We did consider it. In fact, the [class](https://github.com/GoogleCloudPlatform/healthcare/blob/1ce896c9f40d63eb1b6d46c86e647af1d03ca014/imaging/ml/toolkit/hcls_imaging_ml_toolkit/dicom_path.py#L53) that this is based on extends `attr`. During the internal review process @hackermd and I decided to [drop that dependency](https://github.com/MGHComputationalPathology/dicomweb-client/commit/0dc93dc2cdedeae4955d683dee3ba27b9253a67b), because:
- To our knowledge, the standard library does not provide validators for the [UI Value Representation](http://dicom.nema.org/dicom/2013/output/chtml/part05/sect_6.2.html) fields. Perhaps, one could put something together for a sorted sequence of positive integers in the case of frame numbers? While `dataclasses` and `attr` may provide hooks to trigger custom validators, the code savings are minimal.
- Bulk of the effort goes into validating the relationships among the various fields (UIDs, frame numbers, suffix), which the standard library classes are not equipped to help with.
- The functionality offered by the `URI` class goes well beyond that of a data container; with the additional code it requires, carrying along a dependency to simplify a small subset of it did not seem prudent.

That being said, if you think `dataclasses` support additional features that could sufficiently help simplify the implementation in order to justify the dependency trade-off, please let me know.

Reply:
In my mind, a data class does two things.
1. Provide some basic functionality out of the box to simplify construction (eg, hash, eq, and update which you implemented manually)
2. Signal that the class is largely a data container. While this class adds methods, they're largely accessor methods with validation 

To be honest, I don't think it's too critical which you choose, however I think a dataclass is a bit cleaner and signals intent a bit more strongly. I definitely understand not wanting to bring in a new dependency such as attr or pydantic (my personal go to), which is why I suggested a solution in the standard lib.

Reply:
Conceptually, I would consider `URI` a type of string and the class provides methods for constructing, parsing, validating and manipulating string objects.

Another aspect to consider is that `dataclasses` were only introduced in Python 3.7. For now, I would like to continue support for Python 3.6.

Reply:
Not sure I follow the first argument. A dataclass supports user-defined methods. Like I said however, if you're comfortable reimplementing hash, eq, etc., then the only value of a dataclass is signaling intention and the more clear init functionality / declaration.

Conversely, if python 3.6 support is the main blocker, the [dataclass backport](https://pypi.org/project/dataclasses/) should have you covered.

Reply:
The backport would be an option, but I am not a fan of including a dependency just for the implementation of this class. Dataclasses have advantages, but I think the implementation of the class is slick and free of surprises.

Reply:
@ntenenz Please let us know if you have additional thoughts to share, and if we have your permission to close this thread.

Reply:
You should absolutely feel free to proceed as is. My feedback was only a design suggestion and not identifying a flaw in the code itself.

Reply:
Thank you, very much! The discussion helped lend clarity towards the assumptions behind the design choices.